frontend/wasm/src/module/data_segments.rs

```diff
@@ -0,0 +1,266 @@
+//! Module for handling data segment alignment and merging for Miden VM
+//!
+//! Miden VM requires all data segments to be word-aligned (16-byte aligned) because the
+//! hash/unhash instructions operate on words when loaded from the advice provider. This module provides
+//! functionality to ensure all data segments meet this requirement.
+
+/// Represents a data segment with resolved offset
+#[derive(Debug, Clone)]
+pub struct ResolvedDataSegment {
+    /// The absolute offset in linear memory where this segment starts
+    pub offset: u32,
+    /// The initialization data
+    pub data: Vec<u8>,
+    /// The original name/index for debugging purposes
+    pub name: String,
+    /// Whether this is readonly data
+    pub readonly: bool,
+}
+
+impl ResolvedDataSegment {
+    /// Check if this segment is word-aligned (16-byte aligned)
+    pub fn is_word_aligned(&self) -> bool {
+        self.offset % 16 == 0
+    }
+
+    /// Calculate padding needed to align this segment to word boundary
+    pub fn padding_needed(&self) -> u32 {
+        if self.is_word_aligned() {
+            0
+        } else {
+            16 - (self.offset % 16)
+        }
+    }
+
+    /// Align this segment to word boundary by adjusting offset and prepending zeros
+    pub fn align_to_word_boundary(&mut self) {
+        let padding = self.padding_needed();
+        if padding > 0 {
+            // Adjust offset down to nearest word boundary
+            self.offset = self.offset - (self.offset % 16);
```

Comment:
*Question:* Is there any reason why we can't just do `self.offset -= padding`? I would expect that the padding reflects the adjustment to the offset required to achieve word-alignment, and then we avoid effectively recomputing the value of `padding` here.

Reply:
Great question! I found an error in `padding_needed()`. It was calculated for offset adjustment up and not down. I fixed it and use it here in https://github.com/0xMiden/compiler/pull/549/commits/c4a04cd647640baeccdf50f530372a71c5430f51. Thank you!